src/engine/parser/error.rs

```diff
@@ -0,0 +1,190 @@
+// This file is part of rust-web/twig
+//
+// For the copyright and license information, please view the LICENSE
+// file that was distributed with this source code.
+
+//! Typisation of parser errors.
+
+use std::fmt::{self, Display};
+use error::Error;
+use error::{GeneralizeTo, ErrorCode};
+
+use engine::parser::token::TokenErrorCode;
+use engine::parser::job::{self, cursor};
+use engine::parser::token;
+
+pub type ParserError = Error<ParserErrorCode>;
+pub type NodeError = Error<NodeErrorCode>; // todo move somewhere else??
+
+impl GeneralizeTo<ParserErrorCode> for TokenErrorCode {
+    fn generalize(&self) -> ParserErrorCode { ParserErrorCode::TokenError }
+}
+
+#[allow(dead_code)]
+#[derive(Debug)]
+pub enum ParserErrorCode {
+    Unreachable {
+        reason: String,
+        job: job::JobDump,
+    },
+    MissingExtensions,
+    InvalidState {
```

Comment:
No, I suggest to leave `unreachable!` for all the obvious "should not happen" cases: for example, match regexp capture group of 2 elements, the code that handles 3 has to be unreachable, independent on user template, options or any other kind of input.

For all other cases, well, we know the error, so let's just name it.